libpod/in_memory_state.go

```diff
@@ -0,0 +1,279 @@
+}
+
+// NewInMemoryState initializes a new, empty in-memory state
+func NewInMemoryState() (State, error) {
+	state := new(InMemoryState)
+
+	state.pods = make(map[string]*Pod)
+	state.containers = make(map[string]*Container)
+
+	state.podNameIndex = registrar.NewRegistrar()
+	state.ctrNameIndex = registrar.NewRegistrar()
+
+	state.podIDIndex = truncindex.NewTruncIndex([]string{})
+	state.ctrIDIndex = truncindex.NewTruncIndex([]string{})
+
+	return state, nil
+}
+
+// GetContainer retrieves a container from its full ID
+func (s *InMemoryState) GetContainer(id string) (*Container, error) {
+	if id == "" {
+		return nil, ErrEmptyID
+	}
+
+	ctr, ok := s.containers[id]
+	if !ok {
+		return nil, errors.Wrapf(ErrNoSuchCtr, "no container with ID %s found", id)
+	}
+
+	return ctr, nil
+}
+
+// LookupContainer retrieves a container by full ID, unique partial ID, or name
+func (s *InMemoryState) LookupContainer(idOrName string) (*Container, error) {
+	if idOrName == "" {
+		return nil, ErrEmptyID
+	}
+
+	fullID, err := s.ctrNameIndex.Get(idOrName)
+	if err != nil {
+		if err == registrar.ErrNameNotReserved {
+			// What was passed is not a name, assume it's an ID
+			fullID, err = s.ctrIDIndex.Get(idOrName)
+			if err != nil {
+				if err == truncindex.ErrNotExist {
+					return nil, errors.Wrapf(ErrNoSuchCtr, "no container found with name or ID %s", idOrName)
+				} else {
+					return nil, errors.Wrapf(err, "error performing truncindex lookup for ID %s", idOrName)
+				}
+			}
+		} else {
+			return nil, errors.Wrapf(err, "error performing registry lookup for ID %s", idOrName)
+		}
+	}
+
+	ctr, ok := s.containers[fullID]
+	if !ok {
+		// This should never happen
+		return nil, errors.Wrapf(ErrInternal, "mismatch in container ID registry and containers map for ID %s", fullID)
+	}
+
+	return ctr, nil
+}
+
+// HasContainer checks if a container with the given ID is present in the state
+func (s *InMemoryState) HasContainer(id string) (bool, error) {
+	if id == "" {
+		return false, ErrEmptyID
+	}
+
+	_, ok := s.containers[id]
+
+	return ok, nil
+}
+
+// AddContainer adds a container to the state
+// If the container belongs to a pod, the pod must already be present when the
+// container is added, and the container must be present in the pod
+func (s *InMemoryState) AddContainer(ctr *Container) error {
+	if !ctr.valid {
+		return errors.Wrapf(ErrCtrRemoved, "container with ID %s is not valid", ctr.ID())
+	}
+
+	_, ok := s.containers[ctr.ID()]
+	if ok {
+		return errors.Wrapf(ErrCtrExists, "container with ID %s already exists in state", ctr.ID())
+	}
+
+	if ctr.pod != nil {
+		_, ok := s.pods[ctr.pod.ID()]
+		if !ok {
+			return errors.Wrapf(ErrNoSuchPod, "pod %s does not exist, cannot add container %s", ctr.pod.ID(), ctr.ID())
+		}
+
+		hasCtr, err := ctr.pod.HasContainer(ctr.ID())
+		if err != nil {
+			return errors.Wrapf(err, "error checking if container %s is present in pod %s", ctr.ID(), ctr.pod.ID())
+		} else if !hasCtr {
```

Comment:
No need for else.

Reply:
But keeping the else keeps it on the same line. I'll drop it if you really want, but I would prefer keeping it

Reply:
I believe the linter will complain about this.  And I don't see the one less line as being important

Reply:
Linter thinks it's fine, the current issues are just comment format.